src/main/java/org/broadinstitute/hellbender/engine/spark/AssemblyRegionWalkerSpark.java

```diff
@@ -0,0 +1,189 @@
+                .collect(Collectors.toList());
+        List<SimpleInterval> paddedIntervalsForReads =
+                intervals.stream().map(interval -> interval.expandWithinContig(readShardPadding, sequenceDictionary)).collect(Collectors.toList());
+        return paddedIntervalsForReads;
+    }
+
+    /**
+     * Loads assembly regions and the corresponding reference and features into a {@link JavaRDD} for the intervals specified.
+     *
+     * If no intervals were specified, returns all the assembly regions.
+     *
+     * @return all assembly regions as a {@link JavaRDD}, bounded by intervals if specified.
+     */
+    public JavaRDD<Tuple3<AssemblyRegion, ReferenceContext, FeatureContext>> getAssemblyRegions(JavaSparkContext ctx) {
+        SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+        JavaRDD<Shard<GATKRead>> shardedReads = SparkSharder.shard(ctx, getReads(), GATKRead.class, sequenceDictionary, intervalShards, readShardSize, shuffle);
+        Broadcast<ReferenceMultiSource> bReferenceSource = hasReference() ? ctx.broadcast(getReference()) : null;
+        Broadcast<FeatureManager> bFeatureManager = features == null ? null : ctx.broadcast(features);
+        return shardedReads.flatMap(getAssemblyRegionsFunction(bReferenceSource, bFeatureManager, sequenceDictionary, getHeaderForReads(),
+                assemblyRegionEvaluator(), minAssemblyRegionSize, maxAssemblyRegionSize, assemblyRegionPadding, activeProbThreshold, maxProbPropagationDistance));
+    }
+
+    private static FlatMapFunction<Shard<GATKRead>, Tuple3<AssemblyRegion, ReferenceContext, FeatureContext>> getAssemblyRegionsFunction(
+            final Broadcast<ReferenceMultiSource> bReferenceSource,
+            final Broadcast<FeatureManager> bFeatureManager,
+            final SAMSequenceDictionary sequenceDictionary,
+            final SAMFileHeader header,
+            final AssemblyRegionEvaluator evaluator,
+            final int minAssemblyRegionSize,
+            final int maxAssemblyRegionSize,
+            final int assemblyRegionPadding,
+            final double activeProbThreshold,
+            final int maxProbPropagationDistance) {
+        return (FlatMapFunction<Shard<GATKRead>, Tuple3<AssemblyRegion, ReferenceContext, FeatureContext>>) shardedRead -> {
+            SimpleInterval paddedInterval = shardedRead.getPaddedInterval();
+            SimpleInterval assemblyRegionPaddedInterval = paddedInterval.expandWithinContig(assemblyRegionPadding, sequenceDictionary);
+
+            ReferenceDataSource reference = bReferenceSource == null ? null :
+                    new ReferenceMemorySource(bReferenceSource.getValue().getReferenceBases(null, assemblyRegionPaddedInterval), sequenceDictionary);
+            FeatureManager features = bFeatureManager == null ? null : bFeatureManager.getValue();
```

Comment:
It will do with https://github.com/samtools/htsjdk/pull/724. I've tested on a cluster with this change. Any chance of a review there? :)